### PR TITLE
Feat: Add WeChat in Global Config

### PR DESCRIPTION
**Pull request by nutmos**
## Description
Add global config support in the alert manager for WeChat.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
Using the unit test for validation.

## Changelog entry

```release-note
- Add support for API URLs in the global configuration for the alert manager for WeChat
```

Relate #7622

### CODE REVIEW

File: pkg/alertmanager/amcfg_test.go [hunk 1]
```diff
@@ -901,6 +905,360 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
...
+					},
+				},
+			},
+			matcherStrategy: monitoringv1.AlertmanagerConfigMatcherStrategy{
+				Type: "OnNamespace",
+			},
+			wantErr: true,
+		},
+		{
+			name:      "valid global config telegram api url",
```
Comment (AshwinSriram11) [inline pkg/alertmanager/amcfg_test.go]: It might be better to add Telegram test cases in a separate PR.

File: pkg/alertmanager/amcfg_test.go [hunk 2]
```diff
@@ -901,6 +905,360 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
...
+					},
+				},
+			},
+			matcherStrategy: monitoringv1.AlertmanagerConfigMatcherStrategy{
+				Type: "OnNamespace",
+			},
+			wantErr: true,
+		},
+		{
+			name:      "valid global config jira api url",
```
Comment (AshwinSriram11) [inline pkg/alertmanager/amcfg_test.go]: Same for the jira ones

File: pkg/alertmanager/amcfg_test.go [hunk 3]
```diff
@@ -901,6 +905,360 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
...
+					},
+				},
+			},
+			matcherStrategy: monitoringv1.AlertmanagerConfigMatcherStrategy{
+				Type: "OnNamespace",
+			},
+			wantErr: true,
+		},
+		{
+			name:      "invalid global config wechat config missing secret",
```
Comment (AshwinSriram11) [inline pkg/alertmanager/amcfg_test.go]: We can also have a test case for an empty `APICorpID`

Comment (nutmos) [reply pkg/alertmanager/amcfg_test.go]: @AshwinSriram11 I tried writing some test cases but it seems not to produce error as expected. Is it expected to be tested in E2e?

File: pkg/alertmanager/amcfg_test.go (hunk 1 above)
Comment (nutmos) [reply pkg/alertmanager/amcfg_test.go]: @AshwinSriram11 Both Telegram and Jira are duplicated from the existing config that was already merged. I removed all of them in the latest comment already.

File: pkg/alertmanager/amcfg_test.go (hunk 3 above)
Comment (AshwinSriram11) [reply pkg/alertmanager/amcfg_test.go]: I don't think so, maybe we need to validate this manually in the `convertGlobalWeChatConfig` 

Let's wait for @simonpasquier or @slashpai to confirm this

File: pkg/alertmanager/amcfg.go [hunk 4]
```diff
@@ -1766,6 +1770,34 @@ func (cb *ConfigBuilder) convertGlobalJiraConfig(out *globalConfig, in *monitori
...
+
+	if in.APISecret != nil {
+		apiSecret, err := cb.store.GetSecretKey(ctx, crKey.Namespace, *in.APISecret)
+		if err != nil {
+			return fmt.Errorf("failed to get WeChat Secret: %w", err)
+		}
+		out.WeChatAPISecret = apiSecret
+	}
+
+	if in.APICorpID != nil {
```
Comment (AshwinSriram11) [inline pkg/alertmanager/amcfg.go:1858]: Like I mentioned, maybe we need to do the check here but let's wait

**Review by AshwinSriram11 — COMMENTED** [review]
Other than the one test case we discussed, everything looks great :+1:

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 5]
```diff
@@ -557,6 +561,22 @@ type GlobalRocketChatConfig struct {
 	TokenID *v1.SecretKeySelector `json:"tokenID,omitempty"`
 }
 
+type GlobalWeChatConfig struct {
+	// WeChat API URL
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go]:
```suggestion
	// The default WeChat API URL.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 6]
```diff
@@ -557,6 +561,22 @@ type GlobalRocketChatConfig struct {
...
+	// The default value is "https://qyapi.weixin.qq.com/cgi-bin/"
+	// +optional
+	APIURL *URL `json:"apiURL,omitempty"`
+
+	// WeChat API Secret
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go]:
```suggestion
	// The default WeChat API Secret.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 7]
```diff
@@ -557,6 +561,22 @@ type GlobalRocketChatConfig struct {
...
+	// +optional
+	APISecret *v1.SecretKeySelector `json:"apiSecret,omitempty"`
+
+	// WeChat API Corporate ID
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go]:
```suggestion
	// The default WeChat API Corporate ID.
```

File: pkg/alertmanager/amcfg.go [hunk 8]
```diff
@@ -1806,6 +1810,37 @@ func (cb *ConfigBuilder) convertGlobalRocketChatConfig(ctx context.Context, out
...
+		if *in.APICorpID == "" {
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go]: it should be prevented by the kubebuilder validation.

File: pkg/alertmanager/amcfg.go [hunk 9]
```diff
@@ -1806,6 +1810,37 @@ func (cb *ConfigBuilder) convertGlobalRocketChatConfig(ctx context.Context, out
...
 
+func (cb *ConfigBuilder) convertGlobalWeChatConfig(ctx context.Context, out *globalConfig, in *monitoringv1.GlobalWeChatConfig, crKey types.NamespacedName) error {
+	if in == nil {
+		return nil
+	}
+
+	if in.APIURL != nil {
+		u, err := url.Parse(string(*in.APIURL))
+		if err != nil {
+			return fmt.Errorf("parse WeChat API URL: %w", err)
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go]:
```suggestion
			return fmt.Errorf("wechat API URL: %w", err)
```

File: pkg/alertmanager/amcfg.go (hunk 8 above)
Comment (nutmos) [reply pkg/alertmanager/amcfg.go]: I'd like to do so but @AshwinSriram11 mentioned before to have a check logic implemented here as well. I'm not sure what is the practice I should follow.

Comment (AshwinSriram11) [reply pkg/alertmanager/amcfg.go]: Sorry if I misguided you. I was not sure what would be the correct method and asked for @simonpasquier's advice. I would suggest proceeding with the original method

Comment (nutmos) [reply pkg/alertmanager/amcfg.go]: @AshwinSriram11 No problem. I will proceed with the original way.

**Comment by nutmos** [comment]
@slashpai @simonpasquier This is ready to review now.

**Review by simonpasquier — APPROVED** [review]
Thanks!

File: pkg/alertmanager/amcfg_test.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/alertmanager/amcfg_test.go]: Ideally we could test in the e2e tests but the validation rules are simple enough that I wouldn't mandate it. It's required for more complex validations like regex.
